File: Swift/Stanford/ObjectiveC/Matchismo/Matchismo/Matchismo/Model/Deck.m
//
//  Deck.m
//  Matchismo
//
//  Created by Tom Lai on 24/3/2018.
//  Copyright © 2018年 Tom Lai. All rights reserved.
//

#import "Deck.h"

@interface Deck()
@property (strong, nonatomic) NSMutableArray* cards;
@end

@implementation Deck

-(NSMutableArray *) cards
{
    if(!_cards) _cards= [[NSMutableArray alloc] init];
    return _cards;
}

-(void) addCard: (Card*) card atTop: (BOOL) atTop{
    if (atTop){
        [self.cards insertObject:card atIndex:0];
    }
    else {
        [self.cards addObject:card];
    }
}
-(void) addCard: (Card*) card{
    [self addCard:card atTop:NO];
}

-(Card*) drawRandomCard{
    Card *randomCard = nil;
    
    unsigned index =arc4random() % [self.cards count];
    randomCard = self.cards[index];
    [self.cards removeObjectAtIndex:index];
    
    return randomCard;
}


@end

File: Swift/Stanford/Swift/Set/Set/Model/Deck.swift
//
//  Deck.swift
//  Set
//
//  Created by Tom Lai on 31/3/2018.
//  Copyright © 2018年 Tom Lai. All rights reserved.
//

import Foundation

class Deck{
    private(set) var cards :  [Card]()
    
    func addCard(card: Card){
        self.addCard(card: card, atTop: false)
    }
    
    func addCard(card: Card, atTop: Bool){
        if atTop{
            cards.insert(card, at: 0)
        }
        else{
            cards.append(card)
        }
    }
    
    func drawRandomCard -> Card{
        let index = Int(arc4raondom() % self.cards.count)
        randomCards = self.cards[index];
        self.cards.remove(index);
        
        return randomCard;
    }
}
